File: test/unit/map/layout/occupancy_test.exs
defmodule WandererApp.Map.Layout.OccupancyTest do
  use ExUnit.Case, async: true

  alias WandererApp.Map.Layout.Geometry
  alias WandererApp.Map.Layout.Occupancy

  @spacing_x Geometry.spacing_x()
  @spacing_y Geometry.spacing_y()

  describe "claim/3 and release/2" do
    test "a claimed slot is taken and a fresh one is free" do
      occ = Occupancy.claim(Occupancy.new(), "a", {0, 0})

      refute Occupancy.free?(occ, {0, 0})
      assert Occupancy.free?(occ, {@spacing_x, 0})
    end

    test "a neighbour one grid step away in either axis is free" do
      occ = Occupancy.claim(Occupancy.new(), "a", {0, 0})

      assert Occupancy.free?(occ, {@spacing_x, 0})
      assert Occupancy.free?(occ, {0, @spacing_y})
      assert Occupancy.free?(occ, {@spacing_x, @spacing_y})
    end

    test "an id can be re-claimed at a new position" do
      occ =
        Occupancy.new()
        |> Occupancy.claim("a", {0, 0})
        |> Occupancy.claim("a", {@spacing_x, @spacing_y})

      assert Occupancy.free?(occ, {0, 0})
      refute Occupancy.free?(occ, {@spacing_x, @spacing_y})
      assert Occupancy.collides_with(occ, {@spacing_x, @spacing_y}) == ["a"]
    end

    test "release frees the slot and tolerates an unknown id" do
      occ =
        Occupancy.new()
        |> Occupancy.claim("a", {0, 0})
        |> Occupancy.release("a")
        |> Occupancy.release("never-claimed")

      assert Occupancy.free?(occ, {0, 0})
    end

    test "an id claimed before the origin is found" do
      occ = Occupancy.claim(Occupancy.new(), "negative", {-3 * @spacing_x, -3 * @spacing_y})

      refute Occupancy.free?(occ, {-3 * @spacing_x, -3 * @spacing_y})
      assert Occupancy.free?(occ, {-2 * @spacing_x, -3 * @spacing_y})
    end
  end

  describe "3x3 neighbourhood completeness" do
    # The lookup only visits the cells around the query's own cell, so these pin
    # down the boundary: an occupant in the NEXT cell that still reaches back
    # into the query must be found, and one a cell further out cannot overlap at
    # all.

    test "finds an occupant in the next cell right that reaches back" do
      occ = Occupancy.claim(Occupancy.new(), "next-cell", {@spacing_x, 0})

      # Query sits at the end of its cell so it reaches into the next one, where
      # the occupant starts exactly on the cell boundary.
      assert Occupancy.collides_with(occ, {@spacing_x - Geometry.node_w() + 1, 0}) == ["next-cell"]
    end

    test "finds an occupant in the previous cell left that reaches forward" do
      occ = Occupancy.claim(Occupancy.new(), "prev-cell", {-Geometry.node_w() + 1, 0})

      assert Occupancy.collides_with(occ, {0, 0}) == ["prev-cell"]
    end

    test "finds an occupant in the neighbouring row below" do
      # Query at y = spacing - 1 is still in row 0; the occupant starts on row 1
      # and overlaps the query's footprint.
      occ = Occupancy.claim(Occupancy.new(), "below", {0, @spacing_y})

      assert Occupancy.collides_with(occ, {0, @spacing_y - 1}) == ["below"]
    end

    test "does not report an occupant whose cell is out of reach" do
      occ = Occupancy.claim(Occupancy.new(), "right", {@spacing_x + Geometry.node_w(), 0})

      assert Occupancy.collides_with(occ, {0, 0}) == []
    end

    test "refuses a cell smaller than a node" do
      assert_raise ArgumentError, ~r/smaller than the node box/, fn ->
        Occupancy.new(Geometry.node_w() - 1, Geometry.node_h() - 1)
      end
    end
  end

  describe "collides_with/3 options" do
    test "skips ignored ids" do
      occ = Occupancy.claim(Occupancy.new(), "self", {0, 0})
      ignore = MapSet.new(["self"])

      assert Occupancy.collides_with(occ, {0, 0}, ignore_ids: ignore) == []
      assert Occupancy.free?(occ, {0, 0}, ignore_ids: ignore)
    end

    test "reports every colliding id" do
      occ =
        Occupancy.new()
        |> Occupancy.claim("a", {0, 0})
        |> Occupancy.claim("b", {0, 0})

      assert Enum.sort(Occupancy.collides_with(occ, {0, 0})) == ["a", "b"]
    end
  end

  describe "from_positions/2 and from_systems/2" do
    test "indexes every position" do
      occ = Occupancy.from_positions(%{"a" => {0, 0}, "b" => {@spacing_x, 0}})

      refute Occupancy.free?(occ, {0, 0})
      refute Occupancy.free?(occ, {@spacing_x, 0})
      assert Occupancy.free?(occ, {2 * @spacing_x, 0})
    end

    test "omits ignored ids" do
      occ = Occupancy.from_positions(%{"a" => {0, 0}, "b" => {@spacing_x, 0}}, MapSet.new(["a"]))

      assert Occupancy.free?(occ, {0, 0})
      refute Occupancy.free?(occ, {@spacing_x, 0})
    end

    test "indexes systems and skips ones without coordinates" do
      systems = [
        %{solar_system_id: 1, position_x: 0, position_y: 0},
        %{solar_system_id: 2, position_x: nil, position_y: nil},
        %{solar_system_id: 3, position_x: @spacing_x, position_y: 0}
      ]

      occ = Occupancy.from_systems(systems)

      refute Occupancy.free?(occ, {0, 0})
      refute Occupancy.free?(occ, {@spacing_x, 0})
      assert Occupancy.free?(occ, {2 * @spacing_x, 0})
    end

    test "keeps hidden systems indexed so an unhidden node cannot stack" do
      # The R-tree still blocks a hidden system's slot until cleanup runs, so
      # leaving it out here would let a new node land on top of it.
      systems = [%{solar_system_id: 1, position_x: 0, position_y: 0, visible: false}]

      refute Occupancy.free?(Occupancy.from_systems(systems), {0, 0})
    end
  end

  describe "find_free_slot/3" do
    test "returns the desired slot untouched when it is free" do
      {position, status} = Occupancy.find_free_slot(Occupancy.new(), {2 * @spacing_x, 3 * @spacing_y})

      assert status == :ok
      assert position == {2 * @spacing_x, 3 * @spacing_y}
    end

    test "nudges the same column before moving sideways" do
      occ = Occupancy.claim(Occupancy.new(), "taken", {0, 0})

      assert {position, :ok} = Occupancy.find_free_slot(occ, {0, 0})
      # (0,-1) precedes (0,1) in the deterministic order.
      assert position == {0, -@spacing_y}
    end

    test "prefers the desired column, then the other side, then further out" do
      occ =
        Occupancy.new()
        |> Occupancy.claim("centre", {0, 0})
        |> Occupancy.claim("up", {0, -@spacing_y})
        |> Occupancy.claim("down", {0, @spacing_y})

      assert {position, :ok} = Occupancy.find_free_slot(occ, {0, 0})
      # Both vertical neighbours are taken, so it steps to one side rather than
      # jumping two rows up.
      assert position == {-@spacing_x, 0}
    end

    test "is deterministic across calls with the same occupancy" do
      occ =
        Occupancy.new()
        |> Occupancy.claim("a", {0, 0})
        |> Occupancy.claim("b", {@spacing_x, 0})

      assert Occupancy.find_free_slot(occ, {0, 0}) == Occupancy.find_free_slot(occ, {0, 0})
    end

    test "is idempotent: claiming the result makes the next search pick elsewhere" do
      occ = Occupancy.claim(Occupancy.new(), "a", {0, 0})

      {first, :ok} = Occupancy.find_free_slot(occ, {0, 0})
      second_occ = Occupancy.claim(occ, "b", first)
      {second, :ok} = Occupancy.find_free_slot(second_occ, {0, 0})

      assert second != first
      assert Occupancy.free?(second_occ, second)
    end

    test "keeps an off-grid desired slot as-is rather than snapping it" do
      # A layout is anchored wherever the map's nodes are, so snapping to
      # absolute grid multiples would misalign the new node from its neighbours.
      desired = {2 * @spacing_x + 7, 3 * @spacing_y - 4}

      assert {^desired, :ok} = Occupancy.find_free_slot(Occupancy.new(), desired)
    end

    test "keeps off-grid neighbours evenly spaced when it has to move" do
      desired = {2 * @spacing_x + 7, 3 * @spacing_y - 4}
      occ = Occupancy.claim(Occupancy.new(), "taken", desired)

      assert {position, :ok} = Occupancy.find_free_slot(occ, desired)
      assert position == {2 * @spacing_x + 7, 2 * @spacing_y - 4}
    end

    test "honours ignored ids" do
      occ = Occupancy.claim(Occupancy.new(), "self", {0, 0})

      assert {{0, 0}, :ok} =
               Occupancy.find_free_slot(occ, {0, 0}, ignore_ids: MapSet.new(["self"]))
    end

    test "stays within about one screen of the desired slot" do
      {position, :ok} = Occupancy.find_free_slot(Occupancy.new(), {0, 0})

      assert position == {0, 0}
      # An occupied neighbourhood is what makes the radius matter.
      occ = Enum.reduce(0..20, Occupancy.new(), fn i, acc -> Occupancy.claim(acc, i, {i * 5, 0}) end)
      {position, _status} = Occupancy.find_free_slot(occ, {0, 0})
      assert position != {0, 0}
    end

    test "reports failure with a deterministic overflow slot when the area is full" do
      occ = Occupancy.claim(Occupancy.new(), "taken", {0, 0})

      assert {position, :exhausted} = Occupancy.find_free_slot(occ, {0, 0}, max_radius: 0)
      assert position == {@spacing_x, 0}
    end
  end

  describe "find_free_x_offset/4" do
    # `base_shift_x` is the translation that would place the block where wanted;
    # the result is that shift, or a larger one that clears the occupancy.

    test "keeps the requested shift when nothing is in the way" do
      block = %{"a" => {0, 0}, "b" => {@spacing_x, 0}}

      assert {offset, :ok} = Occupancy.find_free_x_offset(Occupancy.new(), block, 4 * @spacing_x)
      assert offset == 4 * @spacing_x
    end

    test "translates the whole block past an occupant, preserving its shape" do
      occ = Occupancy.claim(Occupancy.new(), "blocker", {4 * @spacing_x, 0})
      block = %{"a" => {0, 0}, "b" => {@spacing_x, 0}}

      assert {offset, :ok} = Occupancy.find_free_x_offset(occ, block, 4 * @spacing_x)
      assert offset == 5 * @spacing_x
      # Shape preserved: b stays exactly one column right of a.
      assert block["a"] |> elem(0) |> Kernel.+(offset) == 5 * @spacing_x
      assert block["b"] |> elem(0) |> Kernel.+(offset) == 6 * @spacing_x
    end

    test "reports failure one step past the search range" do
      occ =
        Enum.reduce(0..3, Occupancy.new(), fn i, acc ->
          Occupancy.claim(acc, "filler-#{i}", {(4 + i) * @spacing_x, 0})
        end)

      assert {offset, :exhausted} =
               Occupancy.find_free_x_offset(occ, %{"a" => {0, 0}}, 4 * @spacing_x, max_steps: 3)

      assert offset == 8 * @spacing_x
    end

    test "ignores the ids the caller excludes" do
      occ = Occupancy.claim(Occupancy.new(), "blocker", {4 * @spacing_x, 0})

      assert {offset, :ok} =
               Occupancy.find_free_x_offset(occ, %{"a" => {0, 0}}, 4 * @spacing_x,
                 ignore_ids: MapSet.new(["blocker"])
               )

      assert offset == 4 * @spacing_x
    end

    test "accepts an empty block at the requested shift" do
      assert {offset, :ok} = Occupancy.find_free_x_offset(Occupancy.new(), %{}, 2 * @spacing_x)
      assert offset == 2 * @spacing_x
    end
  end

  describe "block_extent/1" do
    test "returns the horizontal span of a block" do
      assert Occupancy.block_extent(%{"a" => {0, 0}, "b" => {3 * @spacing_x, 0}}) ==
               {0, 3 * @spacing_x}
    end

    test "returns a zero span for an empty block" do
      assert Occupancy.block_extent(%{}) == {0, 0}
    end
  end

  describe "Geometry.overlaps?/2" do
    test "touching edges are adjacent, not stacked" do
      refute Geometry.overlaps?({0, 0}, {@spacing_x, 0})
      refute Geometry.overlaps?({0, 0}, {0, @spacing_y})
    end

    test "any real intersection counts" do
      assert Geometry.overlaps?({0, 0}, {Geometry.node_w() - 1, 0})
      assert Geometry.overlaps?({0, 0}, {0, Geometry.node_h() - 1})
      assert Geometry.overlaps?({0, 0}, {0, 0})
    end
  end
end

File: assets/js/hooks/Mapper/components/map/helpers/occupancy.test.ts
import { NODE_H, NODE_W, SPACING_X, SPACING_Y } from './geometry';
import {
  blockExtent,
  claim,
  collidesWith,
  createOccupancy,
  findFreeSlot,
  findFreeXOffset,
  isFree,
  occupancyOfNodes,
  occupancyOfPositions,
  release,
} from './occupancy';

const at = (x: number, y: number) => ({ x, y });

describe('Occupancy', () => {
  describe('claim / release', () => {
    it('reports a claimed slot as taken and a fresh one as free', () => {
      const occ = createOccupancy();
      claim(occ, 'a', at(0, 0));

      expect(isFree(occ, 0, 0)).toBe(false);
      expect(isFree(occ, SPACING_X, 0)).toBe(true);
    });

    it('treats a neighbour one grid step away in either axis as free', () => {
      const occ = createOccupancy();
      claim(occ, 'a', at(0, 0));

      expect(isFree(occ, SPACING_X, 0)).toBe(true);
      expect(isFree(occ, 0, SPACING_Y)).toBe(true);
      expect(isFree(occ, SPACING_X, SPACING_Y)).toBe(true);
    });

    it('lets an id be re-claimed at a new position', () => {
      const occ = createOccupancy();
      claim(occ, 'a', at(0, 0));
      claim(occ, 'a', at(SPACING_X, SPACING_Y));

      expect(isFree(occ, 0, 0)).toBe(true);
      expect(isFree(occ, SPACING_X, SPACING_Y)).toBe(false);
      expect(collidesWith(occ, SPACING_X, SPACING_Y)).toEqual(['a']);
    });

    it('frees the cell on release and tolerates releasing an unknown id', () => {
      const occ = createOccupancy();
      claim(occ, 'a', at(0, 0));
      release(occ, 'a');
      release(occ, 'never-claimed');

      expect(isFree(occ, 0, 0)).toBe(true);
    });

    it('finds a claim whose cell lies before the origin', () => {
      const occ = createOccupancy();
      claim(occ, 'negative', at(-SPACING_X * 3, -SPACING_Y * 3));

      expect(isFree(occ, -SPACING_X * 3, -SPACING_Y * 3)).toBe(false);
      expect(isFree(occ, -SPACING_X * 2, -SPACING_Y * 3)).toBe(true);
    });
  });

  describe('3x3 neighbourhood completeness', () => {
    // The lookup only visits the cells around the query's own cell, so these
    // pin down the boundary: an occupant in the NEXT cell that still reaches
    // back into the query must be found, and one a cell further out cannot
    // intersect at all.

    it('finds an occupant in the next cell right that reaches back', () => {
      const occ = createOccupancy();
      // Query sits at the end of its cell (x=51 in cell 0) so it reaches into
      // cell 1, where the occupant starts exactly on the cell boundary.
      claim(occ, 'next-cell', at(SPACING_X, 0));

      expect(collidesWith(occ, SPACING_X - NODE_W + 1, 0)).toEqual(['next-cell']);
    });

    it('finds an occupant in the previous cell left that reaches forward', () => {
      const occ = createOccupancy();
      claim(occ, 'prev-cell', at(-NODE_W + 1, 0));

      expect(collidesWith(occ, 0, 0)).toEqual(['prev-cell']);
    });

    it('finds an occupant in the neighbouring row below', () => {
      const occ = createOccupancy();
      // Query at y=74 is still in row 0; the occupant starts on row 1 at y=75
      // and overlaps the query's 34px footprint.
      claim(occ, 'below', at(0, SPACING_Y));

      expect(collidesWith(occ, 0, SPACING_Y - 1)).toEqual(['below']);
    });

    it('does not report an occupant whose cell is out of reach', () => {
      const occ = createOccupancy();
      claim(occ, 'right', at(SPACING_X + NODE_W, 0));

      expect(collidesWith(occ, 0, 0)).toEqual([]);
    });

    it('throws when a cell is smaller than a node', () => {
      const occ = createOccupancy(NODE_W - 1, NODE_H - 1);
      claim(occ, 'a', at(0, 0));

      expect(() => isFree(occ, 0, 0)).toThrow();
    });
  });

  describe('collidesWith options', () => {
    it('skips ignored ids', () => {
      const occ = createOccupancy();
      claim(occ, 'self', at(0, 0));

      expect(collidesWith(occ, 0, 0, { ignoreIds: new Set(['self']) })).toEqual([]);
      expect(isFree(occ, 0, 0, { ignoreIds: new Set(['self']) })).toBe(true);
    });

    it('reports every colliding id', () => {
      const occ = createOccupancy();
      claim(occ, 'a', at(0, 0));
      claim(occ, 'b', at(0, 0));

      expect(collidesWith(occ, 0, 0).sort()).toEqual(['a', 'b']);
    });
  });

  describe('occupancyOfPositions / occupancyOfNodes', () => {
    it('indexes every position', () => {
      const occ = occupancyOfPositions({ a: at(0, 0), b: at(SPACING_X, 0) });

      expect(isFree(occ, 0, 0)).toBe(false);
      expect(isFree(occ, SPACING_X, 0)).toBe(false);
      expect(isFree(occ, SPACING_X * 2, 0)).toBe(true);
    });

    it('omits ignored ids', () => {
      const occ = occupancyOfPositions({ a: at(0, 0), b: at(SPACING_X, 0) }, new Set(['a']));

      expect(isFree(occ, 0, 0)).toBe(true);
      expect(isFree(occ, SPACING_X, 0)).toBe(false);
    });

    it('omits hidden nodes but keeps visible ones', () => {
      const occ = occupancyOfNodes([
        { id: 'hidden', position: at(0, 0), hidden: true },
        { id: 'shown', position: at(SPACING_X, 0) },
      ]);

      expect(isFree(occ, 0, 0)).toBe(true);
      expect(isFree(occ, SPACING_X, 0)).toBe(false);
    });
  });

  describe('findFreeSlot', () => {
    it('returns the desired slot untouched when it is free', () => {
      const occ = createOccupancy();
      const result = findFreeSlot(occ, at(SPACING_X * 2, SPACING_Y * 3));

      expect(result.ok).toBe(true);
      expect(result.position).toEqual(at(SPACING_X * 2, SPACING_Y * 3));
    });

    it('nudges the same column before moving sideways', () => {
      const occ = createOccupancy();
      claim(occ, 'taken', at(0, 0));

      const result = findFreeSlot(occ, at(0, 0));

      expect(result.ok).toBe(true);
      // (0,-1) precedes (0,1) in the deterministic order.
      expect(result.position).toEqual(at(0, -SPACING_Y));
    });

    it('prefers the requested side once the column is blocked too', () => {
      const occ = createOccupancy();
      claim(occ, 'centre', at(0, 0));
      claim(occ, 'up', at(0, -SPACING_Y));
      claim(occ, 'down', at(0, SPACING_Y));

      expect(findFreeSlot(occ, at(0, 0), { prefer: -1 }).position).toEqual(at(-SPACING_X, 0));
      expect(findFreeSlot(occ, at(0, 0), { prefer: 1 }).position).toEqual(at(SPACING_X, 0));
    });

    it('falls back to the other side when the preferred one is taken', () => {
      const occ = createOccupancy();
      claim(occ, 'centre', at(0, 0));
      claim(occ, 'up', at(0, -SPACING_Y));
      claim(occ, 'down', at(0, SPACING_Y));
      claim(occ, 'right', at(SPACING_X, 0));

      const result = findFreeSlot(occ, at(0, 0), { prefer: 1 });

      expect(result.ok).toBe(true);
      expect(result.position).toEqual(at(-SPACING_X, 0));
    });

    it('is deterministic across calls with the same occupancy', () => {
      const occ = createOccupancy();
      claim(occ, 'a', at(0, 0));
      claim(occ, 'b', at(SPACING_X, 0));

      const first = findFreeSlot(occ, at(0, 0));
      const second = findFreeSlot(occ, at(0, 0));

      expect(first.position).toEqual(second.position);
    });

    it('is idempotent: claiming the result makes the next search pick elsewhere', () => {
      const occ = createOccupancy();
      claim(occ, 'a', at(0, 0));

      const first = findFreeSlot(occ, at(0, 0));
      claim(occ, 'b', first.position);
      const second = findFreeSlot(occ, at(0, 0));

      expect(second.position).not.toEqual(first.position);
      expect(isFree(occ, second.position.x, second.position.y)).toBe(true);
    });

    it('search radius stays within about one screen', () => {
      const occ = createOccupancy();
      const result = findFreeSlot(occ, at(0, 0), { maxRadius: 8 });

      expect(result.position.x).toBeGreaterThanOrEqual(-8 * SPACING_X);
      expect(result.position.x).toBeLessThanOrEqual(8 * SPACING_X);
      expect(result.position.y).toBeGreaterThanOrEqual(-8 * SPACING_Y);
      expect(result.position.y).toBeLessThanOrEqual(8 * SPACING_Y);
    });

    it('keeps an off-grid desired slot as-is rather than snapping it', () => {
      const occ = createOccupancy();
      // A subscription layout is anchored at a data coordinate, so snapping to
      // absolute grid multiples would misalign the node from its neighbours.
      const desired = at(SPACING_X * 2 + 7, SPACING_Y * 3 - 4);

      expect(findFreeSlot(occ, desired).position).toEqual(desired);
    });

    it('keeps off-grid neighbours evenly spaced when it has to move', () => {
      const occ = createOccupancy();
      const desired = at(SPACING_X * 2 + 7, SPACING_Y * 3 - 4);
      claim(occ, 'taken', desired);

      const result = findFreeSlot(occ, desired);

      expect(result.ok).toBe(true);
      expect(result.position).toEqual(at(SPACING_X * 2 + 7, SPACING_Y * 2 - 4));
    });

    it('honours ignored ids', () => {
      const occ = createOccupancy();
      claim(occ, 'self', at(0, 0));

      const result = findFreeSlot(occ, at(0, 0), { ignoreIds: new Set(['self']) });

      expect(result.position).toEqual(at(0, 0));
    });

    it('reports failure with a deterministic overflow slot when the area is full', () => {
      const occ = createOccupancy();
      claim(occ, 'taken', at(0, 0));

      const result = findFreeSlot(occ, at(0, 0), { maxRadius: 0, prefer: 1 });

      expect(result.ok).toBe(false);
      expect(result.position).toEqual(at(SPACING_X, 0));
    });
  });

  describe('findFreeXOffset', () => {
    // `baseShiftX` is the translation that would place the block where wanted;
    // the result is that shift, or a larger one that clears the occupancy.
    it('keeps the requested shift when nothing is in the way', () => {
      const occ = createOccupancy();
      const block = { a: at(0, 0), b: at(SPACING_X, 0) };

      const result = findFreeXOffset(occ, block, SPACING_X * 4);

      expect(result.ok).toBe(true);
      expect(result.offsetX).toBe(SPACING_X * 4);
    });

    it('translates the whole block past an occupant, preserving its shape', () => {
      const occ = createOccupancy();
      claim(occ, 'blocker', at(SPACING_X * 4, 0));

      const block = { a: at(0, 0), b: at(SPACING_X, 0) };
      const result = findFreeXOffset(occ, block, SPACING_X * 4);

      expect(result.ok).toBe(true);
      expect(result.offsetX).toBe(SPACING_X * 5);
      // Shape preserved: b stays exactly one column right of a.
      expect(block.a.x + result.offsetX).toBe(SPACING_X * 5);
      expect(block.b.x + result.offsetX).toBe(SPACING_X * 6);
    });

    it('reports failure one step past the search range', () => {
      const occ = createOccupancy();
      for (let i = 0; i <= 3; i++) claim(occ, `filler-${i}`, at(SPACING_X * (4 + i), 0));

      const block = { a: at(0, 0) };
      const result = findFreeXOffset(occ, block, SPACING_X * 4, { maxSteps: 3 });

      expect(result.ok).toBe(false);
      expect(result.offsetX).toBe(SPACING_X * 8);
    });

    it('ignores the ids the caller excludes', () => {
      const occ = createOccupancy();
      claim(occ, 'blocker', at(SPACING_X * 4, 0));

      const block = { a: at(0, 0) };
      const result = findFreeXOffset(occ, block, SPACING_X * 4, { ignoreIds: new Set(['blocker']) });

      expect(result.ok).toBe(true);
      expect(result.offsetX).toBe(SPACING_X * 4);
    });

    it('accepts an empty block at the requested shift', () => {
      const occ = createOccupancy();
      const result = findFreeXOffset(occ, {}, SPACING_X * 2);

      expect(result.ok).toBe(true);
      expect(result.offsetX).toBe(SPACING_X * 2);
    });
  });

  describe('blockExtent', () => {
    it('returns the horizontal span of a block', () => {
      expect(blockExtent({ a: at(0, 0), b: at(SPACING_X * 3, 0) })).toEqual({ minX: 0, maxX: SPACING_X * 3 });
    });

    it('returns a zero span for an empty block', () => {
      expect(blockExtent({})).toEqual({ minX: 0, maxX: 0 });
    });
  });
});
